### ISSUE TITLE
[jvm] automatic conversion of Haxe functions to Java functional interfaces

### ISSUE THREAD

**Issue by Gama11** [HaxeFoundation/haxe#9576]
In Java, any interface with a single function is considered a "functional interface" that can be "instantiated" with  lambda syntax. So for interop purposes, it would be quite nice if the following magically worked:

```haxe
import java.util.function.Function;

function foo(f:Function<Int, Int>) {
	trace(f.apply(3));
}

function main() {
	foo(i -> i * 2);
}
```

(with `foo()` usually being loaded from a `.jar`)

**Comment by Simn** [comment]
I have added `Function` as one of the hardcoded implicit interfaces. I was wondering if we should solve this in a broader manner, but I'm not sure how exactly that would work. Maybe it's more pragmatic to just go through the list of `java.util.function` interfaces and add them all like I did in this commit.

There's also a test case. The example for now requires a `cast` because I have no idea how to satisfy unification here.

**Comment by Gama11** [comment]
Hm, too bad you need a cast there.

>but I'm not sure how exactly that would work

It'd have to be done dynamically for, any single-method-interface I guess?

**Comment by Simn** [comment]
But https://docs.oracle.com/javase/8/docs/api/java/util/function/Function.html isn't even single-method.

**Comment by Gama11** [comment]
The default methods probably don't count.

**Comment by Simn** [comment]
Now that we actually track these default methods there's a chance to fix this. On the generator side, we will have to identify all single-method interfaces and then check each of these against each function class. That shouldn't be very expensive because it is going to fail very early very often.

The other problem is the unification part...

**Comment by Simn** [comment]
Implemented in #11019.
